### PR TITLE
Fix sumoperator

### PR DESCRIPTION
**Pull request by gpescia**
There was a bug in adding different sum-operators which should be removed with this PR. Also it is now possible to compute the kinetic energy for complex wf.

### CODE REVIEW

File: netket/operator/_sumoperators.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ class SumOperator(ContinuousOperator):
 
     def __init__(
         self,
-        *operators: List,
+        operators: List,
```
Comment (PhilipVinc) [inline netket/operator/_sumoperators.py:32]: This is breaking existing code (And you'd be breaking a lot of my internal code) so I would be in principle against this.
Why did you want to do this?

File: netket/operator/_sumoperators.py [hunk 2]
```diff
@@ -85,7 +98,7 @@ def _expect_kernel_batched(
         return sum(result)
 
     def _pack_arguments(self):
-        return [self._coeff * jnp.array(op._pack_arguments()) for op in self._ops]
+        return list(self._coeff)
```
Comment (PhilipVinc) [inline netket/operator/_sumoperators.py:101]: I think this is wrong, because that way you are not passing down the `pack_arguments` of the various terms inside the SumOperator.

File: netket/operator/_sumoperators.py (hunk 1 above)
Comment (gpescia) [reply netket/operator/_sumoperators.py:32]: ah but the operator test are all passing?

File: netket/operator/_sumoperators.py (hunk 2 above)
Comment (PhilipVinc) [inline netket/operator/_sumoperators.py:104]:
```suggestion
        return self._coeff, [op._pack_arguments() for op in self._ops]
```

and then change expect_kernel_batched to
```python
    def _expect_kernel_batched(
        self, logpsi: Callable, params: PyTree, x: Array, data: Optional[PyTree]
    ):
    term_coefficients, term_datas = data

        result = [
            term_coefficients[i] * op._expect_kernel_batched(logpsi, params, x, term_datas[i])
            for i, op in enumerate(self._ops)
        ]
```

Comment (gpescia) [reply netket/operator/_sumoperators.py:104]: I do it in the init of the operator now.

File: netket/operator/_continuous_operator.py [hunk 3]
```diff
@@ -92,6 +92,12 @@ def __mul__(self, other):
...
-            return SumOperator(self, coefficients=other)
+            return SumOperator([self], coefficients=[other])
         else:
             return NotImplemented  # pragma: no cover
+
+    def __sub__(self, other):
+        if isinstance(self, ContinuousOperator) and isinstance(
+            other, ContinuousOperator
+        ):
+            return self + (-1.0) * other
```
Comment (PhilipVinc) [inline netket/operator/_continuous_operator.py:103]:
```suggestion
            return self + ( - other)
```

and implement
```python
    def __neg__(self):
        return -1.0 * self
```
as well?

Comment (gpescia) [reply netket/operator/_continuous_operator.py:103]: makes sense

File: netket/operator/_kinetic.py [hunk 4]
```diff
@@ -20,10 +20,35 @@
 import jax.numpy as jnp
 
 from netket.utils.types import DType, PyTree, Array
+import netket.jax as nkjax
 from netket.hilbert import AbstractHilbert
 from netket.operator import ContinuousOperator
 
 
+def jacrev(f):
```
Comment (PhilipVinc) [inline netket/operator/_kinetic.py:28]: can you remove those changes and move them to another PR?

Comment (gpescia) [reply netket/operator/_kinetic.py:28]: ok

File: netket/operator/_sumoperators.py (hunk 2 above)
Comment (PhilipVinc) [reply netket/operator/_sumoperators.py:104]: yeah... i understand it works for your use-case here, but SumOperator is more general and it's supposed to support more arbitrary operators that might not have a single array in their `_pack arguments`. 

Like, stuff with more than 1 array.

Can you please revert this change? I think if you implement it like I propose below it should still work (and it will be less brittle)

File: netket/operator/_sumoperators.py (hunk 1 above)
Comment (gpescia) [reply netket/operator/_sumoperators.py:32]: ah yes I just replaced with [op1,op2,op3] but no need for that indeed

File: netket/operator/_sumoperators.py (hunk 2 above)
Comment (gpescia) [reply netket/operator/_sumoperators.py]: The coefficient and _pack_arguments is now split. Can you still have a look at the init?

File: netket/operator/_sumoperators.py [hunk 5]
```diff
@@ -40,6 +40,21 @@ def __init__(
...
         """
+        if len(operators) != len(coefficients):
+            raise AssertionError("Each operator needs a coefficient")
+
+        new_operators = []
+        new_coeffs = []
+        for op, c in zip(operators, coefficients):
+            if isinstance(op, SumOperator):
+                new_operators = new_operators + op._ops
+                new_coeffs = new_coeffs + list(c * jnp.array(op._pack_arguments()[0]))
```
Comment (PhilipVinc) [inline netket/operator/_sumoperators.py:51]:
```suggestion
                new_coeffs = new_coeffs + list(c * op._coeff)
```

I think you forgot to update this as well

File: netket/operator/_sumoperators.py [hunk 6]
```diff
@@ -40,6 +40,21 @@ def __init__(
...
+            else:
+                new_operators.append(op)
+                new_coeffs.append(c)
+
+        operators = new_operators
+        coefficients = new_coeffs
```
Comment (PhilipVinc) [inline netket/operator/_sumoperators.py:57]: For (small) performance reasons, it would be better to store the coefficients as a jnp array
I can't write the suggestion here, but you should put it below the `self._dtype=dtype` line 
`self._coeff = jnp.asarray(coefficients, dtype=dtype)`

File: netket/operator/_sumoperators.py [hunk 7]
```diff
@@ -77,15 +94,18 @@ def _expect_kernel(
...
 
         return sum(result)
 
     def _pack_arguments(self):
-        return [self._coeff * jnp.array(op._pack_arguments()) for op in self._ops]
+
+        return self._coeff, [op._pack_arguments() for op in self._ops]
 
     def __repr__(self):
         return f"SumOperator(coefficients={self._pack_arguments()})"
```
Comment (PhilipVinc) [inline netket/operator/_sumoperators.py:111]:
```suggestion
        return f"SumOperator(coefficients={self._coeff}))"
```

File: netket/operator/_sumoperators.py [hunk 8]
```diff
@@ -40,6 +40,21 @@ def __init__(
             coefficients: A coefficient for each ContinuousOperator object
             dtype: Data type of the matrix elements. Defaults to `np.float64`
         """
+        if len(operators) != len(coefficients):
+            raise AssertionError("Each operator needs a coefficient")
+
```
Comment (PhilipVinc) [inline netket/operator/_sumoperators.py:45]: It's a minor convention, but we keep the validation of inputs at the top of the function, so this block should go above here.

```python
        hil = [op.hilbert for op in operators]
          if not all(_ == hil[0] for _ in hil):
            raise NotImplementedError(
                "Cannot add operators on different hilbert spaces"
            )
```

File: netket/operator/_sumoperators.py [hunk 9]
```diff
@@ -40,6 +40,21 @@ def __init__(
             coefficients: A coefficient for each ContinuousOperator object
             dtype: Data type of the matrix elements. Defaults to `np.float64`
         """
+        if len(operators) != len(coefficients):
+            raise AssertionError("Each operator needs a coefficient")
```
Comment (PhilipVinc) [inline netket/operator/_sumoperators.py:44]: I think that for ease of use it would be useful to accept a scalar value for `coefficients` (as documented by the signature) and upscale it to the right length

```suggestion
        if is_scalar(coefficients):
            coefficients = [coefficients for _ in operators]

        if len(operators) != len(coefficients):
            raise AssertionError("Each operator needs a coefficient")
```

you will also need to import
```
from netket.utils import is_scalar
```

File: netket/operator/_potential.py [hunk 10]
```diff
@@ -43,14 +43,14 @@ def __init__(
...
+        self.coefficient = jnp.asarray(coefficient, dtype=dtype)
 
-        super().__init__(hilbert, self.coefficient.dtype)
+        super().__init__(hilbert, dtype)
 
     def _expect_kernel(
         self, logpsi: Callable, params: PyTree, x: Array, coefficient: Optional[PyTree]
     ):
-        return coefficient * self._afun(x)
+        return jnp.array(coefficient) * self._afun(x)
```
Comment (PhilipVinc) [inline netket/operator/_potential.py:53]: you are converting in the initialisation to a jnp array so no need to do it again here
```suggestion
        return coefficient * self._afun(x)
```

File: netket/operator/_continuous_operator.py [hunk 11]
```diff
@@ -78,7 +78,7 @@ def __add__(self, other):
         ):
             from netket.operator import SumOperator
 
-            return SumOperator(self, other)
+            return SumOperator(self, other, coefficients=[1.0, 1.0])
```
Comment (PhilipVinc) [inline netket/operator/_continuous_operator.py:81]: with the changes I proposed below, this should now work
```suggestion
            return SumOperator(self, other)
```

File: netket/operator/_continuous_operator.py [hunk 12]
```diff
@@ -92,6 +92,20 @@ def __mul__(self, other):
         if isinstance(self, ContinuousOperator) and isinstance(other, float):
             from netket.operator import SumOperator
 
-            return SumOperator(self, coefficients=other)
+            return SumOperator(self, coefficients=[other])
```
Comment (PhilipVinc) [inline netket/operator/_continuous_operator.py:95]:
```suggestion
            return SumOperator(self, coefficients=other)
```

File: netket/operator/_continuous_operator.py [hunk 13]
```diff
@@ -95,3 +95,17 @@ def __mul__(self, other):
             return SumOperator(self, coefficients=other)
         else:
             return NotImplemented  # pragma: no cover
+
+    def __sub__(self, other):
+        if isinstance(self, ContinuousOperator) and isinstance(
+            other, ContinuousOperator
+        ):
```
Comment (PhilipVinc) [inline netket/operator/_continuous_operator.py:102]: I mean, this is a continuous operator so self is a continuousoperator, isn't it?
```suggestion
        if isinstance(other, ContinuousOperator):
```

File: netket/operator/_continuous_operator.py [hunk 14]
```diff
@@ -95,3 +95,17 @@ def __mul__(self, other):
...
+            return self + (-other)
+        else:
+            return NotImplemented  # pragma: no cover
+
+    def __neg__(self):
+        if isinstance(self, ContinuousOperator):
+            return -1.0 * self
+        else:
+            return NotImplemented  # pragma: no cover
```
Comment (PhilipVinc) [inline netket/operator/_continuous_operator.py:111]:
```suggestion
    def __neg__(self):
        return -1.0 * self
```

Comment (PhilipVinc) [reply netket/operator/_continuous_operator.py:111]: or am I missing something?

File: netket/operator/_sumoperators.py [hunk 15]
```diff
@@ -40,13 +41,30 @@ def __init__(
...
-
         hil = [op.hilbert for op in operators]
         if not all(_ == hil[0] for _ in hil):
             raise NotImplementedError(
                 "Cannot add operators on different hilbert spaces"
             )
 
+        if is_scalar(coefficients):
+            coefficients = [coefficients for _ in operators]
+        if len(operators) != len(coefficients):
```
Comment (PhilipVinc) [inline netket/operator/_sumoperators.py:52]: just typesetting
```suggestion
            coefficients = [coefficients for _ in operators]

        if len(operators) != len(coefficients):
```

**Review by PhilipVinc — APPROVED** [review]
with those last three nitpicks (please double check)

File: netket/operator/_sumoperators.py [hunk 16]
```diff
@@ -67,8 +86,10 @@ def is_hermitian(self):
     def _expect_kernel(
         self, logpsi: Callable, params: PyTree, x: Array, data: Optional[PyTree]
     ):
+        term_coefficients, term_datas = data
         result = [
-            op._expect_kernel(logpsi, params, x, data[i])
+            term_coefficients[i]
+            * op._expect_kernel_batched(logpsi, params, x, term_datas[i])
```
Comment (PhilipVinc) [inline netket/operator/_sumoperators.py:92]: bug

Comment (PhilipVinc) [inline netket/operator/_sumoperators.py:92]:
```suggestion
            * op._expect_kernel(logpsi, params, x, term_datas[i])
```

**Review by PhilipVinc — CHANGES_REQUESTED** [review]

**Review by PhilipVinc — APPROVED** [review]
